internal/builtinfunctions/functions.go

```diff
@@ -164,7 +166,46 @@ func getFloatToStringFunction() schema.CallableFunction {
 	return funcSchema
 }
 
-// TODO: Webb can add a function called floatToStringAdvanced that allows other float formats
+func getFloatToFormattedStringFunction() schema.CallableFunction {
+	funcSchema, err := schema.NewCallableFunction(
+		"floatToFormattedString",
+		[]schema.Type{
+			schema.NewFloatSchema(nil, nil, nil),
+			schema.NewStringSchema(nil, nil, regexp.MustCompile(`^[beEfFgGxX]$`)),
+			schema.NewIntSchema(nil, nil, nil),
+		},
+		// 'b' format: -ddddp±ddd
+		// 'e' format: -d.dddde±dd
+		// 'E' format: -d.ddddE±dd
+		// 'f' format: -ddd.dddd
+		// 'x' format: -0xd.ddddp±ddd
+		// 'X' format: -0Xd.ddddP±ddd
+		schema.NewStringSchema(
+			nil,
+			nil,
+			regexp.MustCompile(`^-?(?:0[xX])?\d+(?:\.\d*)?(?:[pPeE][-+]/d{2,3})?$`)),
+		false,
+		schema.NewDisplayValue(
+			schema.PointerTo("floatToFormattedString"),
+			schema.PointerTo(
+				"Converts a floating point number to a string according to the "+
+					"specified formatting directive and precision."+
+					" Param 1: the floating point value to convert\n"+
+					" Param 2: the format specifier: 'b', 'e', 'E', 'f', 'g', 'G', 'x', 'X'\n"+
+					" Param 3: the number of digits included in the fraction portion.\n"+
+					" (See https://pkg.go.dev/strconv@go1.22.0#FormatFloat for details.)",
+			),
+			nil,
+		),
+		func(f float64, fmt string, precision int64) string {
+			return strconv.FormatFloat(f, fmt[0], int(precision), 64)
```

Comment:
Your downcast for `precision` made me search for the Go standard library documentation, which, oddly, doesn't specify a type for the `prec` parameter.

But ... if you're going to blindly downcast from `int64` to `int`, which I believe just truncates, should you (paranoiacally?) do a range check to be sure you're not screwing up someone's ridiculously large precision value? I mean, nobody would do that, and, if they did, they've got no right to assume it'd work, but unchecked user errors just bug me. 😆

Reply:
The [docs](https://pkg.go.dev/strconv@go1.22.0#FormatFloat) say:
> func FormatFloat(f [float64](https://pkg.go.dev/builtin#float64), fmt [byte](https://pkg.go.dev/builtin#byte), prec, bitSize [int](https://pkg.go.dev/builtin#int)) [string](https://pkg.go.dev/builtin#string)

I believe that doesn't omit the type for `prec` as much as _elide_ it -- I think it means that both `prec` and `bitSize` have the same type (😝) which is `int`.  (In any case, the compiler was unhappy with me until I added the type conversion.  😢)

While I share your instincts with regard to checking the precision value, I don't know what this code should do if it finds it out of range.  (I _did_ try some very large precision values, and I got some very long strings in return, even though the precision of the underlying representation is only accurate to about 17 digits, and after 54 it just produces zeros; so, if someone were to use a value larger than 4.3 billion, I suspect that they would be **relieved** to have us truncate it....)

Reply:
Yeah, I assumed the implication that Go did things that way, although I didn't investigate the details and it's really bizarre syntax considering that Go loudly praises itself on "left to right" readability in declarations (i.e., putting the shared type at the right rather than the left feels glaringly contradictory).

In any case, yeah, anyone who really wants higher precision in the output is insane; but dropping it silently *still* bugs me. Ah, well. I'm just resolving this.